### ISSUE TITLE
Infinite recursion during compilation

### ISSUE THREAD

**Issue by Warchant** [Naios/fluent.cpp#21]
Hi @Naios 

Our team develops https://github.com/soramitsu/kagome/ in C++17. We decided to use continuable instead of callbacks to simplify our code.

We integrated latest master (42af23f) and started to refactor interfaces.

Me and my team mate (separately) run into the same problem of infinite recursion during compilation. 
We both use latest Mac OS, latest CLion for development, same compiler (`Apple LLVM version 10.0.1 (clang-1001.0.46.4)`).

In my case, it occurred after I added this chain (branch `feature/ref-transport-continuable`):
https://github.com/soramitsu/kagome/blob/feature/ref-transport-continuable/core/libp2p/transport/tcp/tcp_transport.cpp#L29

It looks legit, and I don't see any bugs there, but when you build `tcp_transport` target, it just never finishes (recurses infinitely). At some point I run out of disk space (clangd consumed more than 70GB) and compilation died.

**Steps to reproduce**:
I don't have time to create small self-container repo with repro code, but instead I will just share our project.

```
1. git clone --recursive https://github.com/soramitsu/kagome/
2. cd kagome
3. mkdir build
4. cd build
# this step may take a while, since hunter builds all our dependencies during cmake time
# after it completes, you can remove ~/.hunter
5. cmake .. 
6. make tcp_transport
```

**Comment by Rogiel** [comment]
@Warchant, I ran into the same issue a while back. This is a bug in Clang and it has been already fixed (in Clang).

I remember that I managed to find a workaround for it, but sadly I do not remember what it was. I believe this was fixed in Clang 7.0. You could install Clang 7+ using Homebrew or another package manager and use a newer Clang without this bug. It's a bit sad how slow Apple is when updating their AppleClang with upstream Clang.

**Comment by Naios** [comment]
Thanks @Rogiel for your help on this.
@Warchant could you test whether updating Apple Clang to version 7 fixes your issues then?
Often a workaround for such an issue is to spell out auto types for instance `auto&& promise` -> `cti::promise promise` but this has a negative performance effect.

**Comment by Warchant** [comment]
@Naios 

1. clang-8.0.0_1 - problem persists.
```
clang version 8.0.0 (tags/RELEASE_800/final)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /usr/local/Cellar/llvm/8.0.0_1/bin
```

2. AppleClang 10.0.1 - problem persists
```
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

![image](https://user-images.githubusercontent.com/1867551/59560977-da768900-9022-11e9-9d8a-e741da711864.png)
![image](https://user-images.githubusercontent.com/1867551/59560985-ebbf9580-9022-11e9-9c1f-75204e7ef337.png)
... and grows 3-5 GB every few seconds

**Comment by Warchant** [comment]
Explicit types solved the problem

```diff
conn->resolve(rendpoint.value())
-    .then([conn](auto&& iterator) {
+    .then([conn](TcpConnection::ResolverResultsType iterator) {
       return conn->connect(std::move(iterator));
     })
     .then([self, conn]() {
       return self->upgrader_->upgradeToSecure(conn);
     })
-    .then([self](auto&& secureConn) {
+    .then([self](std::shared_ptr<connection::SecureConnection> secureConn) {
      return self->upgrader_->upgradeToMuxed(std::move(secureConn));
    })
    .then([promise{std::move(promise)}](
-              auto&& capableConn) mutable {
+              std::shared_ptr<connection::CapableConnection> capableConn) mutable {
      return promise.set_value(std::move(capableConn));
    });
```

**Comment by Naios** [comment]
I will take a look into this if I have some spare time. The issue probably isn't severe if this can be fixed through concrete types. Also it might be that this isn't related to this library at all and caused by a compiler bug.

**Comment by honzchen** [comment]
The problem persists even if the types are explicitly specified as follows:

```cpp
cti::make_continuable<std::string>([](cti::promise<std::string> promise) {
    promise.set_value("hello");
}).then([](std::string str) {
    // do not return new continuable
}).then([](int num) {
    
});
```

```cpp
cti::make_continuable<std::string>([](cti::promise<std::string> promise) {
    promise.set_value("hello");
}).then([](int num /* inconsistent type */) {
    
});
```

**Comment by Naios** [comment]
@eldenchen Can you give more details about your affected environment? (compiler and toolchain version)?

**Comment by honzchen** [comment]
```
Apple clang version 11.0.0 (clang-1100.0.33.16)
Target: x86_64-apple-darwin18.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by Naios** [comment]
I've found the problem that caused this issue:

The library triggers a static_assert correctly if a callable can't be called with a subset of its arguments but it seems like this is not enough to stop clang causing the generation
of a 0 - 1 -> std::size_t::max length index sequence which obviously causes infinite work.

I've specialized the code such that the generation of an infinited index sequence is prevented (this happened before only if the code was malformed and a static_assert was triggered previously).

Thanks @eldenchen for the reproducible example.
